tests/gtest/aviftest_helpers.cc

```diff
@@ -75,57 +75,28 @@ AvifImagePtr CreateImage(int width, int height, int depth,
 }
 
 void FillImagePlain(avifImage* image, const uint32_t yuva[4]) {
-  avifPixelFormatInfo info;
-  avifGetPixelFormatInfo(image->yuvFormat, &info);
-
   for (int c = 0; c < 4; c++) {
-    uint8_t* row = (c == AVIF_CHAN_A) ? image->alphaPlane : image->yuvPlanes[c];
-    if (!row) {
-      continue;
-    }
-    const uint32_t row_bytes =
-        (c == AVIF_CHAN_A) ? image->alphaRowBytes : image->yuvRowBytes[c];
-    const uint32_t plane_width =
-        (c == AVIF_CHAN_Y || c == AVIF_CHAN_A)
-            ? image->width
-            : ((image->width + info.chromaShiftX) >> info.chromaShiftX);
-    const uint32_t plane_height =
-        (c == AVIF_CHAN_Y || c == AVIF_CHAN_A)
-            ? image->height
-            : ((image->height + info.chromaShiftY) >> info.chromaShiftY);
+    const uint32_t plane_width = avifImagePlaneWidth(image, c);
+    const uint32_t plane_height = avifImagePlaneHeight(image, c);
     for (uint32_t y = 0; y < plane_height; ++y) {
+      uint8_t* row = avifImagePlaneRow(image, c, y);
```

Comment:
Unless the compiler inlines this `avifImagePlaneRow()` call, the compiler won't be able to perform the common optimization of converting this to the original code (i.e., replacing multiplcation with addition).

I think we just need to provide a `avifImagePlane()` function that returns a pointer to the first row of the plane.

Reply:
I am not really a CS person. According to this Wikipedia page, the compiler optimization I described in the previous comment is called strength reduction:

https://en.wikipedia.org/wiki/Strength_reduction

Reply:
The code I have in mind would call `avifImagePlaneRowBytes()` outside the `y` for loop and save the return value in a local variable, similar to `plane_width` and `plane_height`.

For example, the `CopyImageSamples()` function in tests/gtest/aviftest_helpers.cc would be written as follows (I found that it's very close to the original code):

```
void CopyImageSamples(const avifImage& from, avifImage* to) {
  ...

  for (int c = 0; c < 4; c++) {
    const uint8_t* from_row = avifImagePlane(&from, c);
    uint8_t* to_row = avifImagePlane(to, c);
    assert(!from_row == !to_row);
    if (!from_row) {
      continue;
    }
    const uint32_t from_row_bytes = avifImagePlaneRowBytes(&from, c);
    const uint32_t to_row_bytes = avifImagePlaneRowBytes(to, c);
    const uint32_t plane_width = avifImagePlaneWidth(&from, c);
    const uint32_t plane_height = avifImagePlaneHeight(&from, c);
    for (uint32_t y = 0; y < plane_height; ++y) {
      if (avifImageUsesU16(&from)) {
        std::copy(reinterpret_cast<const uint16_t*>(from_row),
                  reinterpret_cast<const uint16_t*>(from_row) + plane_width,
                  reinterpret_cast<uint16_t*>(to_row));
      } else {
        std::copy(from_row, from_row + plane_width, to_row);
      }
      from_row += from_row_bytes;
      to_row += to_row_bytes;
    }
  }
}
```

Since you wrote the original code and I remember I didn't suggest this structure to you, I am wondering if you now see some problem with this structure. (This structure is commonly used.)

Or do you think `avifImagePlaneRow()` will be handy, for example, when we need to get a pointer to the last row:

```
    const uint32_t plane_height = avifImagePlaneHeight(image, channel);
    uint8_t * last_row = avifImagePlaneRow(image, channel, plane_height - 1);
```

Reply:
> Since you wrote the original code and I remember I didn't suggest this structure to you, I am wondering if you now see some problem with this structure.

I do not see any problem with this structure. It was the only structure available with libavif API at the time. In this PR, I was just trying to shorten it from

```c
  for (int c = 0; c < 4; c++) {
    const uint8_t* from_row = avifImagePlane(from, c);
    uint8_t* to_row = avifImagePlane(to, c);
    const uint32_t from_row_bytes = avifImagePlaneRowBytes(from, c);
    const uint32_t to_row_bytes = avifImagePlaneRowBytes(to, c);
    const uint32_t plane_height = avifImagePlaneHeight(from, c);
    for (uint32_t y = 0; y < plane_height; ++y) {
      // do something with from_row and to_row
      from_row += from_row_bytes;
      to_row += to_row_bytes;
    }
  }
```
to
```c
  for (int c = 0; c < 4; c++) {
    for (uint32_t y = 0; y < avifImagePlaneHeight(from, c); ++y) {
      const uint8_t* from_row = avifImagePlaneRow(from, c, y);
      uint8_t* to_row = avifImagePlaneRow(to, c, y);
      // do something with from_row and to_row
    }
  }
```

but let's keep `avifImagePlane()` for consistency with the commonly used patterns.

> Or do you think `avifImagePlaneRow()` will be handy, for example, when we need to get a pointer to the last row

This is not a very common case. This example was not part of my motivation to introduce `avifImagePlaneRow()`, although `avifImagePlaneRow()` being a superset of `avifImagePlane()`, it could be used for any pattern.

Reply:
Thank you for clarifying your previous comment. Calling `avifImagePlaneRow(from, c, y)` inside the `y` for loop means the code in `avifImagePlaneRow(from, c, y)` that depends only on `from` and `c` is executed repeatedly.